protocol/contracts/core/bera/HoneyPot.sol

```diff
@@ -0,0 +1,121 @@
+pragma solidity ^0.8.4;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+import "@openzeppelin/contracts/utils/cryptography/draft-EIP712.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+
+
+/**
+ * @title HoneyPost Mint Shard Pass
+ *
+ * @dev Gives Puzzle solvers whitelist to enter the honey pot vault
+ *
+ */
+contract HoneyPot is ERC721, EIP712, Ownable, Pausable {
+
+    using Counters for Counters.Counter;
+
+    address public verifier;
+    mapping(address => bool) public minted;
+
+    Counters.Counter private _tokenIdTracker;
+
+    bytes32 public immutable VERIFY_TYPEHASH = keccak256("mint(address account)");
+
+ 
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        address _verifier
+    ) ERC721(_name, _symbol) EIP712(_name, "1") {
+        verifier = _verifier;
+    }
+
+
+    function setVerifier(address _verifier) external onlyOwner {
+        verifier = _verifier;
+    }
+
+    /**
+     * @dev Pause minting of tokens
+     *
+     *  Requirements:
+     *  caller must be admin
+     */
+    function pause() external onlyOwner {
+         _pause();
+    }
+
+    /**
+     * @dev UnPause minting of tokens
+     *
+     *  Requirements:
+     *  caller must be admin
+     */
+    function unpause() external onlyOwner {
+        _unpause();
+    }
+    
+
+    /**
+     * @dev Creates a new token for `msg.sender`. Its token ID will be automatically
+     * assigned. 
+     *
+     * Requirements:
+     *
+     * - User must have completed pussle and got a signature of completion from dapp
+     */
+    function mint(uint8 v, bytes32 r, bytes32 s) public virtual whenNotPaused {
+
+      require(!minted[msg.sender], "Already Minted");
+
+      bytes32 structHash = keccak256(abi.encode(VERIFY_TYPEHASH, msg.sender));
+      bytes32 digest = _hashTypedDataV4(structHash);
+      address signer = ECDSA.recover(digest, v, r, s);
+
+      require(signer == verifier, "invalid signature");
+      _mint(msg.sender, _tokenIdTracker.current());
+      _tokenIdTracker.increment();
+
+      minted[msg.sender] = true;
```

Comment:
update minted[msg.sender] before calling _mint()

Reply:
@nichosystem what re-entrancy vulnerability are you seeing here?

Reply:
_mint() calls _checkOnERC721Received() which then calls onERC721Received() if a contract is minting. A contract could reenter with another mint() during the onERC721Received()

Reply:
hmm @nichosystem _mint() doesn't have _checkOnERC721Received. But I will move it up for sanity sake :)

Reply:
oh nice it's just safeMint, lgtm 👍